rs/registry/transport/src/lib.rs

```diff
@@ -236,26 +241,137 @@ pub fn serialize_get_changes_since_request(version: u64) -> Result<Vec<u8>, Erro
     }
 }
 
+/// This is just a "thin wrapper" around Registry's `get_chunk` method.
+#[automock]
+#[async_trait]
+pub trait GetChunk { // DO NOT MERGE - Rename
+    async fn get_chunk_no_validation( // DO NOT MERGE - Rename
+        &self,
+        content_sha256: &[u8],
+    ) -> Result<Vec<u8>, String>;
+}
+
+/// Returns concatenation of chunks.
+///
+/// Fetches each chunk using get_chunk_with_validation.
+pub async fn fetch_large_value(
+    get_chunk: &impl GetChunk,
+    keys: &LargeValueChunkKeys,
+) -> Result<Vec<u8>, String> {
+    let mut result = vec![];
+    // Chunks could instead be fetched in parallel.
+    for key in &keys.chunk_content_sha256s {
+        let mut chunk_content = get_chunk_with_validation(get_chunk, key).await?;
+        result.append(&mut chunk_content);
+    }
+    Ok(result)
+}
+
+/// Verification is needed because `get_chunk` is a query.
+async fn get_chunk_with_validation(
+    get_chunk: &impl GetChunk,
+    content_sha256: &[u8],
+) -> Result<Vec<u8>, String> {
+    let chunk_content = get_chunk.get_chunk_no_validation(content_sha256).await?;
+
+    // Verify chunk.
+    if Sha256::hash(&chunk_content) != content_sha256 {
+        let len = chunk_content.len();
+        let snippet_len = 20.min(len);
+        return Err(format!(
+            "Chunk content hash does not match: len={}, head={:?}, tail={:?} SHA256={:?}",
+            len,
+            &chunk_content[..snippet_len],
+            &chunk_content[len - snippet_len..len],
+            content_sha256,
+        ));
+    }
+
+    Ok(chunk_content)
+}
+
 /// Deserializes the response obtained from the registry canister for a
 /// get_changes_since() call, from protobuf.
-pub fn deserialize_get_changes_since_response(
+pub async fn deserialize_get_changes_since_response(
```

Comment:
This doesn't seem like the right name for this function anymore, since it is no longer deserializing alone, but is also populating data.

It would seem like we'd want to first deserialize the response, and later process the entries to see which ones need to be further populated.  I would spin out a new function instead of modifying this, and then perhaps modify callers of this function to do the post-processing.  This function can simply do the deserialization